scipy/signal/_upfirdn_apply.pyx

```diff
@@ -65,18 +72,144 @@ def _output_len(Py_ssize_t len_h,
+
+    if data_info.ndim != output_info.ndim:
+        return 1
+    if axis >= data_info.ndim:
+        return 2
+
+    make_temp_data = data_info.strides[axis] != sizeof(DTYPE_t);
+    make_temp_output = output_info.strides[axis] != sizeof(DTYPE_t);
+    if make_temp_data:
+        temp_data = <DTYPE_t*>malloc(data_info.shape[axis] * sizeof(DTYPE_t))
+        if not temp_data:
+            free(temp_data)
+            free(temp_output)
+            return 3
+    if make_temp_output:
+        temp_output = <DTYPE_t*>malloc(output_info.shape[axis] * sizeof(DTYPE_t))
+        if not temp_output:
+            free(temp_data)
+            free(temp_output)
+            return 4
+
+    for i in range(output_info.ndim):
+        if i != axis:
+            num_loops *= output_info.shape[i]
+
+    cdef Py_ssize_t j
+    cdef Py_ssize_t data_offset
+    cdef Py_ssize_t output_offset
+    cdef DTYPE_t* data_row
+    cdef DTYPE_t* output_row
+    cdef Py_ssize_t reduced_idx
+    cdef Py_ssize_t j_rev
+    cdef Py_ssize_t axis_idx
+    cdef DTYPE_t* tmp_ptr = NULL
+    for i in range(num_loops):
+        data_offset = 0
+        output_offset = 0
+        # Calculate offset into linear buffer
+        reduced_idx = i
+        for j in range(output_info.ndim):
+            j_rev = output_info.ndim - 1 - j
+            if j_rev != axis:
+                axis_idx = reduced_idx % output_info.shape[j_rev]
+                reduced_idx /= output_info.shape[j_rev]
+                data_offset += (axis_idx * data_info.strides[j_rev])
+                output_offset += (axis_idx * output_info.strides[j_rev])
+
+        # Copy to temporary data if necessary
+        if make_temp_data:
+            for j in range(data_info.shape[axis]):
+                # Offsets are byte offsets, to need to cast to char and back
+                tmp_ptr = <DTYPE_t *>((<char *> data) + data_offset +
+                    j * data_info.strides[axis])
+                temp_data[j] = tmp_ptr[0]
+
+        # Select temporary or direct output and data
+        if make_temp_data:
+            data_row = temp_data
+        else:
+            data_row = <DTYPE_t *>((<char *>data) + data_offset)
+        if make_temp_output:
+            output_row = temp_output
+            for j in range(output_info.shape[axis]):
+                output_row[j] = 0.0  # initialize as zeros
+        else:
+            output_row = <DTYPE_t *>((<char *>output) + output_offset)
+
+        # call 1D upfirdn
+        _apply_impl(data_row, data_info.shape[axis],
+                    h_trans_flip, len_h, output_row, up, down)
+
+        # Copy from temporary output if necessary
+        if make_temp_output:
+            for j in range(output_info.shape[axis]):
+                # Offsets are byte offsets, to need to cast to char and back
+                tmp_ptr = <DTYPE_t *>((<char *>output) + output_offset +
+                    j * output_info.strides[axis])
+                tmp_ptr[0] = output_row[j]
```

Comment:
> In that case there may be a minor memory overhead as the excess samples produced are all removed once at the end rather than on a row-by-row basis.

Doesn't this still effectively remove the empty samples on a row-by-row basis? It's within the `for i in range(num_loops)` loop, which iterates over the non-`axis` dimensions. It's not like for a 10x100000, axis=0 case you're allocating a 20x100000 array and returning a view; you're allocating a 20-element array and only pulling 10 values from it 100000 times, right?